_posts/2021-08-03-scalable-ets-counters.md

```diff
@@ -0,0 +1,222 @@
+---
+layout: post
+title: Decentralized ETS Counters for Better Scalability
+tags: ETS, erlang term storage, scalability, multicore
+author: Kjell Winblad
+---
+
+A shared [Erlang Term Storage
+(ETS)](https://erlang.org/doc/man/ets.html) table is often an
+excellent place to store data that that is updated and read from
+multiple Erlang processes frequently. ETS provides key-value stores to
+Erlang processes. When the
+[write_concurrency](https://erlang.org/doc/man/ets.html#new-2) option
+is activated, ETS tables use fine-grained locking
+internally. Therefore, a scenario where multiple processes insert and
+remove different items in an ETS table should scale well with the
+number of utilized cores. Unfortunately, in practice, the scalability
+for such scenarios is not yet perfect. However, the ETS table option
+[`decentralized_counters`](https://erlang.org/doc/man/ets.html#new-2)
+(introduced in Erlang/OTP 22 for `ordered_set` tables and in
+Erlang/OTP 23 for the other table types) has made the scalability much
+better. A table with `decentralized_counters` activated uses
+decentralized counters instead of centralized counters to track the
+number of items in the table and the memory consumption. This blog
+post will give you a better understanding of when one should activate
+the `decentralized_counters` option and how the decentralized counters
+work.
+
+## Scalability with Decentralized ETS Counters
+
+The following figure shows the throughput (operations/second) archived
+when processes are doing inserts (`ets:insert/2`) and deletes
+(`ets:delete/2`) to an ETS table of the `set` type on a machine with
+64 hardware threads both when `decentralized_counters` option is
+activated and when it is deactivated. The table types `bag` and
+`duplicate_bag` have similar scalability behavior as their
+implementation is based on the same hash table.
+
+![alt text](/images/ets_scalable_counters/bench_set_50_ins_50_del_nospread.png "Throughput of inserts and deletes on a table of type set with and without the decentralized_counters activated")
+
+The following figure shows the results for the same benchmark but with
+a table of type `ordered_set`:
+
+![alt text](/images/ets_scalable_counters/bench_ordset_50_ins_50_del_nospread.png "Throughput of inserts and deletes on a table of type ordered_set with and without the decentralized_counters activated")
+
+The interested reader can find more benchmark results, benchmark
+details, and the benchmark machine
+[here](http://winsh.me/ets_catree_benchmark/decent_ctrs_hash.html). The
```

Comment:
It is considered bad practice used the word "here" as link text. I think you can shorten the sentence to "The interested reader can find more information about the benchmark." and anchor the link on "more information about the benchmark".

Reply:
Fixed in b8fb1b3. Please have a look to see if you think it is better now.